bigtable/bttest/inmem.go

```diff
@@ -1242,13 +1242,23 @@ func (t *table) gc() {
 		return
 	}
 
+	// It isn't clear whether it's safe to delete within the iterator, so we do
+	// not
+	var toDelete []btree.Item
 	t.rows.Ascend(func(i btree.Item) bool {
 		r := i.(*row)
 		r.mu.Lock()
 		r.gc(rules)
+		if len(r.families) == 0 {
+			toDelete = append(toDelete, i)
+		}
 		r.mu.Unlock()
 		return true
 	})
+
+	for _, i := range toDelete {
```

Comment:
Since the emulator is supposed to be able to handle concurrent operations, it looks like this triggers a race condition (which was flagged by the CI):

```
==================
WARNING: DATA RACE
Write at 0x00c00004c090 by goroutine 18:
  github.com/google/btree.(*BTree).deleteItem()
      /go/pkg/mod/github.com/google/btree@v1.0.1/btree.go:731 +0x144
  github.com/google/btree.(*BTree).Delete()
      /go/pkg/mod/github.com/google/btree@v1.0.1/btree.go:712 +0x43b
  cloud.google.com/go/bigtable/bttest.(*table).gc()
      /tmpfs/src/google-cloud-go/bigtable/bttest/inmem.go:1260 +0x3f6
  cloud.google.com/go/bigtable/bttest.TestConcurrentMutationsReadModifyAndGC.func5()
      /tmpfs/src/google-cloud-go/bigtable/bttest/inmem_test.go:125 +0x6c

Previous read at 0x00c00004c090 by goroutine 40:
  github.com/google/btree.(*BTree).Get()
      /go/pkg/mod/github.com/google/btree@v1.0.1/btree.go:819 +0x144
  cloud.google.com/go/bigtable/bttest.(*table).mutableRow()
      /tmpfs/src/google-cloud-go/bigtable/bttest/inmem.go:1211 +0xfd
  cloud.google.com/go/bigtable/bttest.(*server).MutateRow()
      /tmpfs/src/google-cloud-go/bigtable/bttest/inmem.go:793 +0x1bd
  cloud.google.com/go/bigtable/bttest.TestConcurrentMutationsReadModifyAndGC.func3()
      /tmpfs/src/google-cloud-go/bigtable/bttest/inmem_test.go:109 +0x249

Goroutine 18 (running) created at:
  cloud.google.com/go/bigtable/bttest.TestConcurrentMutationsReadModifyAndGC()
      /tmpfs/src/google-cloud-go/bigtable/bttest/inmem_test.go:123 +0xc8a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Goroutine 40 (running) created at:
  cloud.google.com/go/bigtable/bttest.TestConcurrentMutationsReadModifyAndGC()
      /tmpfs/src/google-cloud-go/bigtable/bttest/inmem_test.go:101 +0xbcc
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202
==================
```

@crwilcox @kolea2 @igorbernstein2  any thoughts on how this could be accomplished without creating a race condition? Or thoughts on other workarounds that would make sense here?

Reply:
Whoops! Need to hold the table lock I think. Sorry, forgot to run with -race. Will fix.